sklearn/utils/_encode.py

```diff
@@ -27,13 +28,85 @@ def _unique(values, *, return_inverse=False):
     if values.dtype == object:
         return _unique_python(values, return_inverse=return_inverse)
     # numerical
-    return np.unique(values, return_inverse=return_inverse)
+    out = np.unique(values, return_inverse=return_inverse)
+
+    if return_inverse:
+        uniques, inverse = out
+    else:
+        uniques = out
+
+    # np.unique will have duplicate missing values at the end of `uniques`
+    # here we clip the nans and remove it from uniques
+    if uniques.size and is_scalar_nan(uniques[-1]):
+        nan_idx = np.searchsorted(uniques, np.nan)
+        uniques = uniques[:nan_idx + 1]
+        if return_inverse:
+            inverse[inverse > nan_idx] = nan_idx
+
+    if return_inverse:
+        return uniques, inverse
+    return uniques
+
+
+def _split_missing(items, none_is_missing=True):
+    """Split items into a set and a list of missing values. If none_is_missing
+    is True, then None and nan are considered missing. If none_is_missing
+    is False, then only nans are considered missing.
+    """
+    def is_missing(value):
+        if none_is_missing:
+            return value is None or is_scalar_nan(value)
+        return is_scalar_nan(value)
+
+    # Return items without missing items
+    missing_values = [value for value in items if is_missing(value)]
+
+    if not missing_values:
+        return items, []
+
+    # Enforces an order where None always comes first
+    if None in missing_values:
+        if len(missing_values) == 1:
+            output_missing_values = [None]
+        else:
+            output_missing_values = [None, np.nan]
+    else:
+        output_missing_values = [np.nan]
+
+    # create set without the missing values
+    output = set(value for value in items if not is_missing(value))
```

Comment:
I think set difference still works and is more succinct
```suggestion
    missing_values = {value for value in items if is_missing(value)}
    output = items - missing_values

    # Enforces an order where None always comes first
    if None in missing_values:
        if len(missing_values) == 1:
            output_missing_values = [None]
        else:
            output_missing_values = [None, np.nan]
    elif missing_values:
        output_missing_values = [np.nan]
    else:
        output_missing_values = []
```

Reply:
Set different does not work with `float('nan')`:

```
values = [float('nan'), float('nan'), 'a', 'b', 'b', 'b']
uniques = set(values)
uniques - set([float('nan')])
# {'a', 'b', nan, nan}
```

Reply:
I don't think that's quite accurate. Rather, set difference for NaN's will work by object identity rather than value comparison. So as long as the NaNs you are subtracting are the same objects you have in the other operand, it should work fine.

Reply:
Ah, that makes sense. The PR updated with your suggestion.